File: Java_Joc_Windows_Android/AndroidStudioProject/core/src/base/menu/AnimatedComposedStiledVisual.java
package base.menu;

import com.badlogic.gdx.Gdx;

import java.util.Map;

import base.menu.util.IMenuAnimationListener;
import base.menu.util.IMenuItemConfig_DynamicProperties;
import base.menu.util.MenuItemConfig;
import base.menu.util.MenuItemConfig_SmartItem;
import base.menu.util.StateChangeDurations;
import local.profile.config.ProfileDataIDs;
import local.app.AppLocal;
import base.visual.IContainer2D;
import base.visual.IDynamicPropertiesObject;

// Each menu or panel extends this class it.
public abstract class AnimatedComposedStiledVisual extends ComposedStiledVisual {

    private StateChangeDurations[] easeDurationsSList;

	protected String startStateID = null;
    protected String endStateID = null;
    private float animationDurationS = 0f;
    private float stateCompletionPercent = 1f;

    protected IMenuAnimationListener menuAnimationListener;

    public AnimatedComposedStiledVisual(AppLocal app, MenuItemConfig[] menuItemConfigs, StateChangeDurations[] easeDurationsSList) {
        super(app, menuItemConfigs);

        this.easeDurationsSList = easeDurationsSList;
    }

    public boolean isActive() {
        if(endStateID.contains("_in_")) {
            return true;
        }

        return false;
    }

	public String getEndStateID() {
		return endStateID;
	}

    protected final float getStateCompletionPercent() {
        return stateCompletionPercent;
    }

	public boolean isAnimating() {
		if(stateCompletionPercent < 1) {
			return true;
		}

		return false;
	}

    public boolean isActiveOrAnimating() {
        return isActive() || isAnimating();
    }

    protected void activeStateChangedEvent() {
        //void
    }

    protected void gotoShowState(String endStateID, boolean doAnimation, IMenuAnimationListener menuAnimationListener, boolean restartAnimation) {
		setShowState(this.endStateID, endStateID, doAnimation, menuAnimationListener, restartAnimation);
	}

    private StateChangeDurations getFirstValidEaseDurationConfig(String startStateID, String endStateID) {
        for(int i=0;i<easeDurationsSList.length;i++) {
            StateChangeDurations valueEntry = easeDurationsSList[i];

            boolean isStartStateCheckPassed = checkIsStatePassed(startStateID, valueEntry.startComposedStateID);
            boolean isEndStateCheckPassed = checkIsStatePassed(endStateID, valueEntry.endComposedStateID);

            if(isStartStateCheckPassed && isEndStateCheckPassed) {
                return valueEntry;
            }
        }

        return null;
    }

    protected void setShowState(String startStateID, String endStateID, boolean doAnimation, IMenuAnimationListener menuAnimationListener, boolean restartAnimation) {
		this.menuAnimationListener = menuAnimationListener;

		if(!restartAnimation && (this.startStateID.equals(startStateID) && this.endStateID.equals(endStateID))) {
			return ;
		}

        boolean oldActiveState = isActive();

		this.startStateID = startStateID;
        this.endStateID = endStateID;

        StateChangeDurations easeDurationConfig = getFirstValidEaseDurationConfig(startStateID, endStateID);
        if(easeDurationConfig == null) {
            app.errorHandler.handleAppCriticalError("Error. 1.", "Error.1.");

            throw new Error("Error. 1.");
        }
        this.animationDurationS = easeDurationConfig.durationS;

		if(doAnimation) {
			stateCompletionPercent = 0;
		}else {
			stateCompletionPercent = 1;
		}

        if(oldActiveState != isActive()) {
            activeStateChangedEvent();
        }

        // Also update absolute values because some absolute values may use other absolute values that are manually modified when the menu state change is requested.
		updateStateData(true, true);

        if(stateCompletionPercent >= 1) {
            if(menuAnimationListener != null) {
                menuAnimationListener.animationFinished(this);

                menuAnimationListener = null;

                animationFinishedEvent();
            }
        }
	}

    protected boolean isVisible() {
        return visualsHolder.getVisible();
    }

    protected void isOnDisplayListChanged() {
        //void
    }

    @Override
    public void doFullVisualsSync() {
        //super.doFullVisualsSync();

        updateStateData(true, true);
    }

    @Override
    public void create(Map<String, Object> createConfigMap) {
        startStateID = computeStateFlags("_out_toIn_");
        endStateID = computeStateFlags("_out_toIn_");

        super.create(createConfigMap);
    }

    @Override
    protected void createVisuals(Map<String, Object> createConfigMap) {
        super.createVisuals(createConfigMap);

        visualsHolder = (IContainer2D)getItem(new String[]{"_root", "visualsHolder"});

        visualsHolder.setVisible(false);
    }

    private void updateStateData(boolean syncAbsoluteValues, boolean syncAnimatingValues) {
        if (isActive() && !visualsHolder.getVisible()) {
            visualsHolder.setVisible(true);

            isOnDisplayListChanged();
        } else if (!isActive() && stateCompletionPercent >= 1 && visualsHolder.getVisible()) {
            visualsHolder.setVisible(false);

            isOnDisplayListChanged();
        }

	    if (!visualsHolder.getVisible()) {
		    return;
	    }

	    syncVisuals(syncAbsoluteValues, syncAnimatingValues, startStateID, endStateID, stateCompletionPercent);
    }

    public boolean isOnDisplayList() {
        if(!visualsHolder.getVisible()) {
            return false;
        }

        return true;
    }

	public String getBaseStateFromVariation(String stateID) {
		if(stateID != null) {
			if (stateID.contains("_in_")) {
				return "_in_";
			} else if (stateID.contains("_out_")) {
				return "_out_";
			}
		}

		Gdx.app.debug("Error", "Unhandled animated menu state fxId.");

		return null;
	}

    public void appLoop(float deltaS, int deltaMS) {
        super.appLoop(deltaS, deltaMS);

        if(stateCompletionPercent < 1) {
            stateCompletionPercent = Math.min(1f, stateCompletionPercent + deltaS * (1f / animationDurationS));

            updateStateData(false, true);

            if(stateCompletionPercent >= 1) {
                if(menuAnimationListener != null) {
                    menuAnimationListener.animationFinished(this);

                    menuAnimationListener = null;
                }

                animationFinishedEvent();
            }
        }
    }

    protected void animationFinishedEvent() {
        if (isActive()) {
            String inState = computeInState();
            if(!getEndStateID().equals(inState)) {
                gotoShowState(inState, true, null, false);
            }
        }
    }

    protected final String computeInState() {
        return computeStateFlags("_in_");
    }

    // baseState: "_in_" / "_out_" / "_out_toIn_";
    protected String computeStateFlags(String baseState) {
        String state = baseState;

        if(!app.profileData.getDataWithID_boolean(ProfileDataIDs.IS_ADS_DISABLED)) {
            state = state + "adsOn_";
        }else {
            state = state + "adsOff_";
        }

        return state;
    }

    protected final void updateShowStateIfChanged(boolean doAnimation, IMenuAnimationListener menuAnimationListener) {
        if(!isActive() || isAnimating()) {
            return ;
        }

        String stateID = computeInState();
        if(!doAnimation || !endStateID.equals(stateID)) {
            gotoShowState(stateID, doAnimation, menuAnimationListener, false);
        }
    }

    @Override
    protected void visualItemCreated(Object item, MenuItemConfig itemConfig, IComposedVisual itemsMap, IComposedVisual itemsMapLayer) {
        // Local code include code from super method.
        // super.visualItemCreated(item, itemConfig, itemsMap, itemsMapLayer);

        if(item instanceof IDynamicPropertiesObject && itemConfig instanceof IMenuItemConfig_DynamicProperties) {
            syncVisualItem(
                    (IDynamicPropertiesObject)item, ((IMenuItemConfig_DynamicProperties) itemConfig).getItemProperties(),
                    true, true,
                    startStateID, endStateID,
                    0f,
                    app, itemsMap, itemsMapLayer,
                    getDataMap()
            );
        }

        if(item instanceof ComposedStiledVisual && item instanceof IDynamicPropertiesObject) {
            ((ComposedStiledVisual) item).createVisuals(createConfigMap);
        }

        if(item instanceof ComposedStiledVisual) {
            ((ComposedStiledVisual)item).doFullVisualsSync();
        }

        if(item instanceof AnimatedComposedStiledVisual && itemConfig instanceof MenuItemConfig_SmartItem) {
            MenuItemConfig_SmartItem cItemConfig = (MenuItemConfig_SmartItem)itemConfig;

            if(cItemConfig.initialBaseStateId != null) {
                String newState = ((AnimatedComposedStiledVisual)item).computeStateFlags(cItemConfig.initialBaseStateId);
                ((AnimatedComposedStiledVisual)item).gotoShowState(newState, false, null, false);
            }
        }
    }
}

File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/menu/AnimatedComposedStiledVisual.cpp
#include "AnimatedComposedStiledVisual.h"
#include <base/exceptions/LogicException.h>
#include <base/app/config/AppConfig.h>
#include <base/math/Math.h>
#include <base/menu/util/IMenuItemConfig_DynamicProperties.h>
#include <base/menu/util/MenuItemConfig_SmartItem.h>
#include <base/util/StringUtil.h>
#include <base/screen/ScreenUtil.h>
#include <base/app/IAppMetrics.h>

AnimatedComposedStiledVisual::AnimatedComposedStiledVisual(IAppMetrics* appMetrics, IAppAssets* appAssets, ArrayList<MenuItemConfig*>* itemConfigs, ArrayList<StateChangeDurations*>* easeDurationsSList)
    : super(appMetrics, appAssets, itemConfigs),
      easeDurationsSList(easeDurationsSList)
{
    screenListener.cb_onActiveScreenMetricsChange = std::bind(&AnimatedComposedStiledVisual::onActiveScreenMetricsChange, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4);
}

void AnimatedComposedStiledVisual::createPost() {
    super::createPost();

    base::ScreenUtil::screen->addListener(&screenListener, 1);
}

void AnimatedComposedStiledVisual::onActiveScreenMetricsChange(
	int resolutionX, int resolutionY,
	float ppiX, float ppiY)
{
	isDirty_viewportMetrics = true;
}

bool AnimatedComposedStiledVisual::isActive() {
    if(StringUtil::contains(endStateID, "_in_")) {
        return true;
    }

    return false;
}

bool AnimatedComposedStiledVisual::isShowing() {
    if(StringUtil::contains(startStateID, "_out_") && StringUtil::contains(endStateID, "_in_")) {
        return true;
    }

    return false;
}

bool AnimatedComposedStiledVisual::isHiding() {
    if(StringUtil::contains(startStateID, "_in_") && StringUtil::contains(endStateID, "_out_")) {
        return true;
    }

    return false;
}

std::string& AnimatedComposedStiledVisual::getEndStateID() {
    return endStateID;
}


std::string AnimatedComposedStiledVisual::getManualStateFlags() {
	return manualStateFlags;
}

void AnimatedComposedStiledVisual::setManualStateFlags(std::string& manualStateFlags, bool doAnimation) {
	this->manualStateFlags = manualStateFlags;

	updateShowStateIfChanged(doAnimation, nullptr);
}


float AnimatedComposedStiledVisual::getStateCompletionPercent() {
    return stateCompletionPercent;
}

void AnimatedComposedStiledVisual::setIsDirty_property() {
	isDirty_property = true;
}

bool AnimatedComposedStiledVisual::isAnimating() {
    if(stateCompletionPercent < 1) {
        return true;
    }

    return false;
}

bool AnimatedComposedStiledVisual::isActiveOrAnimating() {
    return isActive() || isAnimating();
}

void AnimatedComposedStiledVisual::onSelectedStateChanged() {
    //void
}

void AnimatedComposedStiledVisual::gotoShowState(std::string endStateID, bool doAnimation, IAnimationListener* menuAnimationListener, bool restartAnimation) {
    setShowState(this->endStateID, endStateID, doAnimation, menuAnimationListener, restartAnimation);
}

StateChangeDurations* AnimatedComposedStiledVisual::getFirstValidEaseDurationConfig(std::string& startStateID, std::string& endStateID) {
    for(int i=0;i<easeDurationsSList->count();i++) {
        StateChangeDurations* valueEntry = easeDurationsSList->getDirect(i);

        bool isStartStateCheckPassed = checkIsStatePassed(startStateID, valueEntry->startComposedStateID);
        bool isEndStateCheckPassed = checkIsStatePassed(endStateID, valueEntry->endComposedStateID);

        if(isStartStateCheckPassed && isEndStateCheckPassed) {
            return valueEntry;
        }
    }

    throw LogicException(LOC);
}

void AnimatedComposedStiledVisual::setShowState(std::string startStateID, std::string& endStateID, bool doAnimation, IAnimationListener* menuAnimationListener, bool restartAnimation) {
    this->menuAnimationListener = menuAnimationListener;

    if(!restartAnimation && (this->startStateID == startStateID && this->endStateID == endStateID)) {
        return ;
    }

    bool oldActiveState = isActive();

    this->startStateID = startStateID;
    this->endStateID = endStateID;

    StateChangeDurations* easeDurationConfig = getFirstValidEaseDurationConfig(startStateID, endStateID);
    if(easeDurationConfig == nullptr) {
        throw LogicException(LOC);
    }
    this->animationDurationS = easeDurationConfig->durationS;

    if(doAnimation && easeDurationConfig->durationS > 0.0f) {
        stateCompletionPercent = 0.0f;
    }else {
        stateCompletionPercent = 1.0f;
    }

    if(oldActiveState != isActive()) {
        onSelectedStateChanged();
    }

    // Also update absolute values because some absolute values may use other absolute values that are manually modified when the menu state change is requested.
    // Example of problem: Manually setting a visual2D (maybe even a dynamic visual2D) property (eg: width) and using that property for another visual2D.
    updateStateData(true, true);

    if(stateCompletionPercent >= 1) {
        if(menuAnimationListener != nullptr) {
            menuAnimationListener->animationFinished(this);

            menuAnimationListener = nullptr;

            animationFinishedEvent();
        }
    }
}

bool AnimatedComposedStiledVisual::isVisible() {
    return visualsHolder->getIsVisible();
}

void AnimatedComposedStiledVisual::isOnDisplayListChanged() {
    //void
}

void AnimatedComposedStiledVisual::doFullVisualsSync() {
    // Don't call the super function because this does the same thing but also includes animation data.
    //super::doFullVisualsSync();

    updateStateData(true, true);
}

void AnimatedComposedStiledVisual::putDataMapValue(std::string& key, std::shared_ptr<base::IWrappedValue> val, bool doAnimatingVisualsSync) {
    super::putDataMapValue(key, val, doAnimatingVisualsSync);

    if(doAnimatingVisualsSync) {
        updateStateData(false, true);
    }
}

void AnimatedComposedStiledVisual::createFramework() {
    super::createFramework();

    // These values need to be initialised early on.
    startStateID = computeStateFlags("_out_toIn_");
    endStateID = computeStateFlags("_out_toIn_");
}

void AnimatedComposedStiledVisual::createVisuals(Map1D<std::string, void*>* createConfigMap) {
    super::createVisuals(createConfigMap);

    visualsHolder = getItemOptionalAs<IContainer2D*>("_root.visualsHolder");

    visualsHolder->setIsVisible(false);
}

void AnimatedComposedStiledVisual::updateStateData(bool syncAbsoluteValues, bool syncAnimatingValues) {
    if (isActive() && !visualsHolder->getIsVisible()) {
        visualsHolder->setIsVisible(true);

        isOnDisplayListChanged();
    } else if (!isActive() && stateCompletionPercent >= 1 && visualsHolder->getIsVisible()) {
        visualsHolder->setIsVisible(false);

        isOnDisplayListChanged();
    }

    if (!visualsHolder->getIsVisible()) {
        return;
    }

    syncVisuals(syncAbsoluteValues, syncAnimatingValues, startStateID, endStateID, stateCompletionPercent);
}

bool AnimatedComposedStiledVisual::isOnDisplayList() {
    if(!visualsHolder->getIsVisible()) {
        return false;
    }

    return true;
}

std::string AnimatedComposedStiledVisual::getBaseStateFromVariation(std::string& stateID) {
    if (StringUtil::contains(stateID, "_in_")) {
        return "_in_";
    } else if (StringUtil::contains(stateID, "_out_")) {
        return "_out_";
    }else {
        throw LogicException(LOC);
    }
}

void AnimatedComposedStiledVisual::visualLoop(float deltaS, int deltaMS) {
    super::visualLoop(deltaS, deltaMS);
	
    if(stateCompletionPercent < 1) {
        if(isDirty_viewportMetrics || isDirty_property) {
			isDirty_viewportMetrics = false;
			isDirty_property = false;

            updateStateData(true, false);
        }

        stateCompletionPercent = Math::min(1.0f, stateCompletionPercent + deltaS * (1.0f / animationDurationS));

        updateStateData(false, true);

        if(stateCompletionPercent >= 1) {
            if(menuAnimationListener != nullptr) {
                menuAnimationListener->animationFinished(this);

                if (getIsDisposed()) {
                    // The above call dispatched an event and the processor instantly disposed this view.
                    /// View disposal not allowed while code still runs in it, need to use some delayed disposal logic.,
                    throw LogicException(LOC);
                }

                menuAnimationListener = nullptr;
            }

            animationFinishedEvent();
        }
    }else {
        if(isDirty_viewportMetrics || isDirty_property) {
			isDirty_viewportMetrics = false;
			isDirty_property = false;

            updateStateData(true, true);
        }
    }
}

void AnimatedComposedStiledVisual::visualLoopB(float deltaS, int deltaMS) {
	super::visualLoopB(deltaS, deltaMS);

	// This is used so smart_children won't have any update delay (eg. on window resize).
	if (isDirty_viewportMetrics || isDirty_property) {
		isDirty_viewportMetrics = false;
		isDirty_property = false;

		updateStateData(true, true);
	}
}

void AnimatedComposedStiledVisual::animationFinishedEvent() {
    if (isActive()) {
        std::string inState = computeStateFlags("_in_");
        if(getEndStateID() != inState) {
            gotoShowState(inState, true, nullptr, false);
        }
    }
}

// baseState: "_in_" / "_out_" / "_out_toIn_";
std::string AnimatedComposedStiledVisual::computeStateFlags(std::string baseState) {
    std::string state = baseState + manualStateFlags;

    if(appMetrics->getScreen()->getWindowDiagonalInches() < AppConfig::MEDIUM_DEVICE_DIAGONAL_INCHES_MIN) {
        state = state + "SmallDevice_";
        state = state + "Mobile_";
    }else {
        state = state + "Tablet_";
        if(appMetrics->getScreen()->getWindowDiagonalInches() < AppConfig::LARGE_DEVICE_DIAGONAL_INCHES_MIN) {
            state = state + "MediumDevice_";
        }else {
            state = state + "LargeDevice_";
        }
    }

    if(appMetrics->getScreen()->getWindowSizeInches()->x >= appMetrics->getScreen()->getWindowSizeInches()->y) {
        state = state + "Landscape_";
    }else {
        state = state + "Portrait_";
    }

    /*if(!app.profileData.getDataWithID_boolean(ProfileDataIDs.IS_ADS_DISABLED)) {
        state = state + "adsOn_";
    }else {
        state = state + "adsOff_";
    }*/


    return state;
}

void AnimatedComposedStiledVisual::updateShowStateIfChanged(bool doAnimation, IAnimationListener* menuAnimationListener) {
    if(!isActive() || isAnimating()) {
        return ;
    }

    std::string stateID = computeStateFlags("_in_");
    if(!doAnimation || endStateID != stateID) {
        gotoShowState(stateID, doAnimation, menuAnimationListener, false);
    }
}

void AnimatedComposedStiledVisual::visualItemCreated(IGenericDrawable* item, MenuItemConfig* itemConfig, IComposedVisual* itemsMap, IComposedVisual* itemsMapLayer) {
    // Local code include code from super method.
    // super::visualItemCreated(item, itemConfig, itemsMap, itemsMapLayer);

    if(dynamic_cast<IDynamicPropertiesObject2D*>(item) && dynamic_cast<IMenuItemConfig_DynamicProperties*>(itemConfig)) {
        syncVisualItem(
            dynamic_cast<IDynamicPropertiesObject2D*>(item), (dynamic_cast<IMenuItemConfig_DynamicProperties*>(itemConfig))->getItemProperties(),
            true, true,
            startStateID, endStateID,
            0.0f,
            itemsMap, itemsMapLayer,
            getDataMap(),
            appMetrics
        );
    }

    if(dynamic_cast<ComposedStiledVisual*>(item) && dynamic_cast<IDynamicPropertiesObject2D*>(item)) {
        (dynamic_cast<ComposedStiledVisual*>(item))->createVisuals(createConfigMap);
    }

    if(dynamic_cast<ComposedStiledVisual*>(item)) {
        (dynamic_cast<ComposedStiledVisual*>(item))->doFullVisualsSync();
    }

    if(dynamic_cast<AnimatedComposedStiledVisual*>(item) && dynamic_cast<MenuItemConfig_SmartItem*>(itemConfig)) {
        MenuItemConfig_SmartItem* cItemConfig = dynamic_cast<MenuItemConfig_SmartItem*>(itemConfig);

        std::string newState = (dynamic_cast<AnimatedComposedStiledVisual*>(item))->computeStateFlags(cItemConfig->initialBaseStateId);

        (dynamic_cast<AnimatedComposedStiledVisual*>(item))->gotoShowState(newState, false, nullptr, false);
    }
}

void AnimatedComposedStiledVisual::disposeMain() {
    if(screenListener.getIsListening()) {
        base::ScreenUtil::screen->removeListener(&screenListener, 1);
    }

    super::disposeMain();
}

AnimatedComposedStiledVisual::~AnimatedComposedStiledVisual() {
    //void
}
